File: internal/uninstall/remnants.go
package uninstall

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/huaquanghan/mu/internal/utils"
)

// knownAliases maps package names to their actual folder name in ~/.config etc.
var knownAliases = map[string]string{
	"code":                 "Code",
	"google-chrome-stable": "google-chrome",
	"google-chrome":        "google-chrome",
	"chromium-browser":     "chromium",
	"vscode":               "Code",
	"slack-desktop":        "Slack",
	"discord":              "discord",
	"spotify-client":       "spotify",
}

// FindRemnants returns existing remnant directories for a package.
func FindRemnants(name string) []string {
	home, err := os.UserHomeDir()
	if err != nil {
		return nil
	}
	_ = home // used implicitly via XDG helpers which call os.UserHomeDir internally

	// Collect candidate names (original + alias)
	names := []string{name}
	if alias, ok := knownAliases[name]; ok && alias != name {
		names = append(names, alias)
	}

	var found []string
	seen := map[string]bool{}

	for _, n := range names {
		candidates := []string{
			filepath.Join(utils.XDGConfigHome(), n),
			filepath.Join(utils.XDGDataHome(), n),
			filepath.Join(utils.XDGCacheHome(), n),
			filepath.Join("/var/lib", n),
		}
		for _, p := range candidates {
			if seen[p] {
				continue
			}
			seen[p] = true
			if utils.PathExists(p) {
				found = append(found, p)
			}
		}
	}
	return found
}

// RemnantSize returns the total size in bytes of user-owned remnant dirs.
// Skips /var/lib paths (would need sudo).
func RemnantSize(paths []string) int64 {
	var total int64
	for _, p := range paths {
		if strings.HasPrefix(p, "/var/") {
			continue // skip root-owned
		}
		size, err := utils.DirSize(p)
		if err == nil {
			total += size
		}
	}
	return total
}

File: src/uninstall/remnants.rs
//! `remnants.go` — leftover file detection for removed packages: the
//! `knownAliases` name map, `FindRemnants` (the per-name candidate scan over
//! the XDG homes and `/var/lib`), and `RemnantSize` (user-owned remnant
//! sizing, `/var/` skipped).

use std::collections::HashSet;
use std::path::{Path, PathBuf};

use crate::paths;
use crate::size;

use super::Deps;

/// `knownAliases` — package name → the folder name it actually uses in
/// `~/.config` etc.
pub(crate) const KNOWN_ALIASES: &[(&str, &str)] = &[
    ("code", "Code"),
    ("google-chrome-stable", "google-chrome"),
    ("google-chrome", "google-chrome"),
    ("chromium-browser", "chromium"),
    ("vscode", "Code"),
    ("slack-desktop", "Slack"),
    ("discord", "discord"),
    ("spotify-client", "spotify"),
];

/// `knownAliases[name]` lookup.
pub(crate) fn known_alias(name: &str) -> Option<&'static str> {
    KNOWN_ALIASES
        .iter()
        .find(|(k, _)| *k == name)
        .map(|(_, v)| *v)
}

/// `FindRemnants` — production wrapper reading the real environment.
pub fn find_remnants(name: &str) -> Vec<String> {
    find_remnants_in(&Deps::real(), name)
}

/// `FindRemnants` — existing remnant directories for a package under
/// `$XDG_CONFIG_HOME`, `$XDG_DATA_HOME`, `$XDG_CACHE_HOME`, and `/var/lib`,
/// for the package name and its alias. Returns nothing when the home
/// directory cannot be resolved (Go's `os.UserHomeDir` error → nil).
pub(crate) fn find_remnants_in(deps: &Deps, name: &str) -> Vec<String> {
    if deps.home.as_os_str().is_empty() {
        return Vec::new();
    }

    // Collect candidate names (original + alias).
    let mut names = vec![name.to_string()];
    if let Some(alias) = known_alias(name)
        && alias != name
    {
        names.push(alias.to_string());
    }

    let mut found: Vec<String> = Vec::new();
    let mut seen: HashSet<PathBuf> = HashSet::new();

    for n in &names {
        let candidates = [
            deps.config_home.join(n),
            deps.data_home.join(n),
            deps.cache_home.join(n),
            Path::new("/var/lib").join(n),
        ];
        for p in candidates {
            if !seen.insert(p.clone()) {
                continue;
            }
            if paths::path_exists(&p) {
                found.push(p.to_string_lossy().into_owned());
            }
        }
    }
    found
}

/// `RemnantSize` — total size in bytes of user-owned remnant dirs.
/// `/var/` paths are skipped (would need sudo); `DirSize` errors skip the
/// entry (Go adds the size only when `err == nil`, dropping partial totals).
pub fn remnant_size(paths: &[String]) -> i64 {
    let mut total: i64 = 0;
    for p in paths {
        if p.starts_with("/var/") {
            continue; // skip root-owned
        }
        let (size, err) = size::dir_size(Path::new(p));
        if err.is_none() {
            total += size as i64;
        }
    }
    total
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::runner::FakeRunner;
    use crate::uninstall::test_support::{deps_for_test, tempdir};
    use std::fs;
    use std::sync::Arc;

    // remnants_test.go: TestFindRemnants_FindsExistingDirs — Go overrides
    // the XDG dirs via env; the port injects them through Deps.
    #[test]
    fn find_remnants_finds_existing_dirs() {
        let tmp = tempdir("remnants");
        let deps = deps_for_test(&tmp, Arc::new(FakeRunner::new()));

        // Create a fake remnant dir (Go: tmp/config/testpkg).
        let config_dir = tmp.join("config").join("testpkg");
        fs::create_dir_all(&config_dir).unwrap();

        let deps = Deps {
            config_home: tmp.join("config"),
            data_home: tmp.join("data"),
            cache_home: tmp.join("cache"),
            ..deps
        };
        let remnants = find_remnants_in(&deps, "testpkg");
        let config_dir = config_dir.to_string_lossy().into_owned();
        assert!(
            remnants.contains(&config_dir),
            "expected {config_dir} in remnants, got {remnants:?}"
        );
        fs::remove_dir_all(&tmp).ok();
    }

    // remnants_test.go: TestRemnantSize_SkipsVarLib
    #[test]
    fn remnant_size_skips_var_lib() {
        let paths = vec![
            "/var/lib/somepackage".to_string(),
            "/tmp/mu-test-remnant".to_string(),
        ];
        // /var/lib should be skipped, /tmp path doesn't exist so also 0.
        assert_eq!(remnant_size(&paths), 0);
    }
}
